haskell/test/SMT/Model/Proposal/QuorumThreshold.hs

```diff
@@ -0,0 +1,78 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module SMT.Model.Proposal.QuorumThreshold
```

Comment:
Everything in this module in particular should be rewritten, I think.

The current implementation is straight up a copy of the original LIGO one, but it doesn't need to be.

The only thing necessary here is for `updateQuorum` to keep the right signature: `Natural -> ModelT ()`, but all the calculations should be different.

For example:
- `onePlusMaxChangePercent` exists only to save costs in LIGO, we don't need it here.
- do we need functions like `fmul`, `fadd`, etc? Can't we implement a `Num QuorumFraction` instead?

@sras can probably better help you point out how these can be rewritten without the considerations that we had to keep in LIGO

Reply:
@pasqu4le I am wondering how much we can differ from the LIGO implementation becuse we using a similar data structure for storage that we use in LIGO (We had to do it so that the compiled contract can be loaded using our testing infra). 

For example, here itself, if we had a model that stored QuorumThreshold as a double, instead of an Integer numerator, then we could have very different (and hopefully simpler) implementation of the QuorumThreshold update procedure that is being implemented here.

Reply:
@rinn7e So in that line of thought, we might be able to convert the integer numerator from the storage to an equivalent `Double` before starting the calculation, and implement the logic in Haskell using the `Double` values, and convert it back to Integer numerator value before storing it in the storage.

Reply:
@sras going too far off with types may be quite brittle to handle, I think.

I mean, we could use `Double`, but then we'd have to consider that _each operation_ in the source code has a fixed precision, rounding the result of a `Double` at the end will probably not exactly match the result (which is expected).

However even in the same types we can reformulate quite a bit.

IIRC here we changed the original calculations to equivalent ones for the purpose of saving gas cost, we don't need to do that in the Haskell module. I think we need to do this and as such have a check that the two are equivalent and working as expected.

Same about the specific order of calculations and type. We may use `Integer` instead of `Natural`s in the original code only as a cost saving option, if we don't need to do so here I suggest we modify the Haskell model to no do that.

Reply:
Well, I thought we only have a small number of operations to calculate the new quorum value that the required precision could be preserved by using a `Double`. But if it is not, we can try something like the `Decimal` type.

But in general, I think it would be better to go from a simplest model that could possibly work, and then add more complexity to the model as we require to match with the contract in terms or precision or something else. Because I think that way, we ll have a better chance of catching more obvious bugs (because the implementation can be radically simple/different).

(Don't mean to give you guys a hard time, and just sharing my thoughts. Please feel free to ignore)

Reply:
Related to this discussion, I'm not quite sure what the final solution is. To conclude, I should only update where it is necessary: Example: removing `onePlusMaxChangePercent`, remove fmul, fadd, and `Num QuorumFraction`... Let me know if my understanding is correct or not.

Reply:
@rinn7e I think what @pasqu4le was suggesting was to convert the quorum threshold into an Integer at the begining of the calculation, and then compute the new quorum using this Integer value as the old quorum. 

The point is that it could simplify some calculations, (so we won't need fmul, fadd functions) and in the end, we will have an Integer value of the new quorum. If everything works well, then we should have a value that is >= 0. Then we can just convert it to a Natural value update the storage.

Also, if you look at this issue description https://github.com/tqtezos/baseDAO/issues/207, you can see the calculations to compute the new quorum. So ideally we should implement the logic without referring the current ligo/haskell implementation, and only using the formula from the issue.

Reply:
@pasqu4le @sras I've updated `updateQuorum` implementation now, this should be ready for re-review.